src/cpp/src/rag/text_embedding_pipeline.cpp

```diff
@@ -0,0 +1,290 @@
+    };
+
+    std::vector<EmbeddingResult> wait_embed() {
+        m_request.wait();
+
+        // [batch_size, hidden_size]
+        const Tensor last_hidden_state = m_request.get_tensor("last_hidden_state");
+
+        return to_embedding_result(last_hidden_state);
+    };
+
+    std::vector<std::string> format_texts(const std::vector<std::string>& texts) {
+        if (!m_config.embed_instruction) {
+            return texts;
+        }
+
+        std::vector<std::string> formatted;
+        formatted.reserve(texts.size());
+
+        for (auto& text : texts) {
+            formatted.emplace_back(*m_config.embed_instruction + text);
+        }
+        return formatted;
+    }
+
+    std::string format_query(const std::string& text) {
+        if (!m_config.query_instruction) {
+            return text;
+        }
+
+        return *m_config.query_instruction + text;
+    }
+
+    std::vector<EmbeddingResult> to_embedding_result(const Tensor& last_hidden_state) {
+        const auto last_hidden_state_data = last_hidden_state.data<float>();
+
+        std::vector<EmbeddingResult> result;
+        const auto shape = last_hidden_state.get_shape();
+
+        const size_t batch_size = shape[0];
+        const size_t hidden_size = shape[1];
+
+        for (size_t batch = 0; batch < batch_size; batch++) {
+            const auto batch_offset = batch * hidden_size;
+            const auto batch_data = last_hidden_state_data + batch_offset;
+            const std::vector<float> batch_result(batch_data, batch_data + hidden_size);
+            result.push_back(batch_result);
+        }
+
+        return result;
+    }
+
+    std::shared_ptr<Model> apply_postprocessing(std::shared_ptr<Model> model, const Config& config) {
+        ov::preprocess::PrePostProcessor processor(model);
+
+        processor.output().postprocess().custom([this, model, &config](const ov::Output<ov::Node>& node) {
+            if (config.pooling_type == PoolingType::CLS) {
+                return get_cls_pooling_op(node);
+            } else if (config.pooling_type == PoolingType::MEAN) {
+                return get_mean_pooling_op(model, node);
+            }
+
+            OPENVINO_THROW("Pooling type is not supported");
+        });
+
+        if (config.normalize) {
+            processor.output().postprocess().custom([](const ov::Output<ov::Node>& node) {
+                auto axis_const = std::make_shared<op::v0::Constant>(ov::element::i32, ov::Shape{1}, std::vector{1});
+                return std::make_shared<op::v0::NormalizeL2>(node, axis_const, 1e-12, op::EpsMode::ADD);
+            });
+        }
+
+        return processor.build();
+    }
+
+    /**
+     * CLS pooling slices first element from seq_length dimension
+     * [batch_size, seq_length, hidden_size] -> [batch_size, seq_length[0], hidden_size]
+     * [10, 5, 768] -> [10, 768]
+     */
+    std::shared_ptr<op::Op> get_cls_pooling_op(const ov::Output<ov::Node>& last_hidden_state_node) {
```

Comment:
as I see, these post-processing functions don't use class state. Should they be separate free functions, defined in unnamed namespace?

Reply:
Yeah, makes sense, moved to unnamed namespace.